server/src/main/java/io/unitycatalog/server/persist/utils/UriUtils.java

```diff
@@ -0,0 +1,245 @@
+      File file = new File(modelStorageRoot);
+      modelStorageRoot = file.toURI().toString();
+    }
+    // Check if the modelStorageRoot ends with a slash and remove it if it does
+    while (modelStorageRoot.endsWith("/")) {
+      modelStorageRoot = modelStorageRoot.substring(0, modelStorageRoot.length() - 1);
+    }
+
+    return modelStorageRoot;
+  }
+
+  private static String getModelDirectoryURI(String entityFullName) {
+    return getModelStorageRoot() + "/" + entityFullName.replace(".", "/");
+  }
+
+  public static String getModelStorageLocation(String catalogId, String schemaId, String modelId) {
+    return getModelDirectoryURI(catalogId + "." + schemaId + ".models." + modelId);
+  }
+
+  public static String getModelVersionStorageLocation(
+      String catalogId, String schemaId, String modelId, String versionId) {
+    return getModelDirectoryURI(
+        catalogId + "." + schemaId + ".models." + modelId + ".versions." + versionId);
+  }
+
+  public static String createStorageLocationPath(String uri) {
+    return updateDirectoryFromUri(uri, Operation.CREATE, Optional.empty()).toString();
+  }
+
+  public static String deleteStorageLocationPath(String uri) {
+    return updateDirectoryFromUri(uri, Operation.DELETE, Optional.empty()).toString();
+  }
+
+  private static URI updateDirectoryFromUri(
+      String uri,
+      Operation op,
+      Optional<GenerateTemporaryModelVersionCredentialsResponse> credentials) {
+    URI parsedUri = URI.create(uri);
+    validateURI(parsedUri);
+    if (!parsedUri.getScheme().equals("file") && credentials.isEmpty()) {
+      throw new BaseException(
+          ErrorCode.INVALID_ARGUMENT, "Empty credentials passed for a non-file based URI");
+    }
+    try {
+      if (parsedUri.getScheme().equals("file")) {
+        return updateLocalDirectory(parsedUri, op);
+      } else if (parsedUri.getScheme().equals("s3")
+          && credentials.get().getAwsTempCredentials() != null) {
+        return updateS3Directory(parsedUri, op, credentials.get().getAwsTempCredentials());
+      } else if (parsedUri.getScheme().equals("gc")
+          && credentials.get().getGcpOauthToken() != null) {
+        return updateGcDirectory(parsedUri, op, credentials.get().getGcpOauthToken());
+      } else if (parsedUri.getScheme().equals("abfs")
+          && credentials.get().getAzureUserDelegationSas() != null) {
+        return updateAbsDirectory(parsedUri, op, credentials.get().getAzureUserDelegationSas());
+      }
+    } catch (Exception e) {
+      throw new BaseException(
+          ErrorCode.INTERNAL,
+          "Error attempting to "
+              + op.name()
+              + " URI "
+              + parsedUri.toString()
+              + ": "
+              + e.getMessage());
+    }
+    throw new BaseException(
+        ErrorCode.INVALID_ARGUMENT, "Unknown scheme detected: " + parsedUri.getScheme());
+  }
+
+  private static boolean isValidURI(String uri) {
+    try {
```

Comment:
one thing missing here is we are not checking for malicious paths for example file://../../b which can lead to us giving clients access to storage outside storage_root. 

`URI normalized = uri.normalize();
    if (!normalized.getPath().startsWith(uri.getPath())) {
      throw new BaseException(ErrorCode.INVALID_ARGUMENT, "Normalization failed: " + uri.getPath());
    }`

Reply:
I can add some checks here for that.  I think the biggest worry would be for the File based approach that we allow users to relative directory themselves into some real danger.

Reply:
I think this check is to avert the same danger.

Reply:
That said, I don't think users ever get to specify the storage location in their calls to UC OSS.  It's a path we fully provide for them on entity creation.  So unless users are really doing something they shouldnt be doing (manually editing the database), not sure how that could happen here.

Reply:
Yep.  If we're in control and only we will be setting this, then I think we don't need to check for bad paths.

Reply:
We'll need that check once we decide to merge FileUtils with this file. but that's a TODO.